### PR TITLE
[InstCombine] Fold fabs over selects

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch folds fabs over select if it is beneficial. I also tried other interger/fp intrinsics. Only handling fabs shows benefit to some real-world applications.

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/intrinsic-select.ll [hunk 1]
```diff
@@ -280,3 +280,30 @@ entry:
...
+; CHECK-NEXT:    [[CMP:%.*]] = fcmp oeq double [[ABS1]], 0x7FF0000000000000
+; CHECK-NEXT:    [[ABS2:%.*]] = select i1 [[CMP]], double 0.000000e+00, double [[ABS1]]
+; CHECK-NEXT:    ret double [[ABS2]]
+;
+  %abs1 = call double @llvm.fabs.f64(double %a)
+  %cmp = fcmp oeq double %abs1, 0x7FF0000000000000
+  %sel = select i1 %cmp, double -0.000000e+00, double %abs1
+  %abs2 = call double @llvm.fabs.f64(double %sel)
+  ret double %abs2
+}
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/intrinsic-select.ll:309]: Test vectors and flag preservation

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/intrinsic-select.ll:309]: Do we handle fneg? Does the fneg+fabs combo fold the same?

Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/intrinsic-select.ll:322]: > Test vectors and flag preservation

Done. Alive2: https://alive2.llvm.org/ce/z/ijP-me

**Review by arsenm — APPROVED** [review]
This is probably worse if the constant is some weird constant expression, but that's rare.

We should do the same for fneg next.

**Pull request merged**